2015/presentations/condemned_to_reinvent_sql/text.md

```diff
@@ -0,0 +1,342 @@
+functions, appropriately named `Func`. Unfortunately, it doesn't support
+```sql
+select aggregate_function(column order by another_column)
+```
+
+syntax, which is vital for our calculations - grades need to be fitted in order
+in which they were given (otherwise the whole "trend" doesn't make any sense).
+This means we need to extend it a little by modifying `Func.template`
+attribute, like this:
+
+```python
+class LinearFit(Func):
+    contains_aggregate = True
+    function = 'linear_fit'
+    template = '%(function)s(%(expressions)s order by %(ordering)s)'
+
+    def __init__(self, expression, ordering, **extra):
+        super(LinearFit, self).__init__(
+            expression,
+            output_field=ArrayField(models.FloatField()),
+            **extra)
+        self.ordering = self._parse_expressions(ordering)[0]
+
+    def resolve_expression(self, *args, **kwargs):
+        c = super(Func, self).resolve_expression(*args, **kwargs)
+        c.ordering = c.ordering.resolve_expression(*args, **kwargs)
+        return c
+
+    def as_sql(self, compiler, connection, function=None, template=None):
+        ordering_sql, ordering_params = compiler.compile(self.ordering)
+        self.extra['ordering'] = ordering_sql
+
+        return super(LinearFit, self).as_sql(compiler, connection, function, template)
+
+    def get_group_by_cols(self):
+        return []
+```
+
+This seems to do the job (note that grades were given randomly, so trends are
+mosly flat)!
+
+```python
+>>> for i in User.objects.annotate(grade_trend=LinearFit('grade__grade', 'grade__date')):
+...     print i.username, i.grade_trend
+...
+john [-0.0017908297019, 3.77233748271]
+mary [-0.000803833399885, 3.63772475795]
+barbara [-0.0018932620358, 3.65124481328]
+peter [-0.00281602111148, 3.87818118949]
+thomas [0.000325526484835, 3.48609958506]
+kevin [0.00103517422177, 3.3887966805]
+betty [-0.000701401065991, 3.54215076072]
+april [5.72926613309e-05, 3.53482019364]
+bob [-0.000128474452681, 3.49868603043]
+sean [0.000461379537839, 3.54069847856]
+denise [0.000466587961597, 3.36507607192]
+```
+
+## Summary
+
+SQL is often considered an "ugly duckling" in the stack, and most of developers
+avoid it like a plague. True, the syntax isn't the prettiest and sometimes it
+works in a counter-intuitive way. On the other hand, you can achieve amazing
+results with very little amount of rather readable code.
+
+The best part is that you don't need to know upfront what kind of reporting
+you're going to do. Most of the time, you can just design your models as usual
+and write your views/aggregates/whatnot right when you need them.
+
+I would like to encourage everyone dealing with databases to dig a bit deeper
+into the relational world and stop relying exclusively on the ORM. The two
+examples are just a tip of an iceberg, there is *much* more in SQL than you
+think.
+
+PS: Just be sure to use a real relational DBMS. MySQL can't do any of the above.
```

Comment:
As noted above, I think this information is important and it should be placed at the beginning of the article.